### PR TITLE
Avoid auto linkage if user uses cmake [16387]

### PR DESCRIPTION
**Pull request by richiware**

### CODE REVIEW

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/151/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Mac/79/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/143/)

File: src/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -49,10 +49,19 @@ elseif(NOT EPROSIMA_INSTALLER)
...
+        )
 
-    target_compile_definitions(${PROJECT_NAME} PRIVATE ${PROJECT_NAME_UPPER}_SOURCE)
+    target_compile_definitions(${PROJECT_NAME}
+        PRIVATE
+        ${PROJECT_NAME_UPPER}_SOURCE
+	INTERFACE
+        $<$<BOOL:${WIN32}>:${PROJECT_NAME_UPPER}_NO_LIB>
+        PUBLIC
+	$<$<STREQUAL:$<TARGET_PROPERTY:${PROJECT_NAME},TYPE>,SHARED_LIBRARY>:${PROJECT_NAME_UPPER}_DYN_LINK>
```
Comment (jsantiago-eProsima) [inline src/cpp/CMakeLists.txt:63]: Maybe this one should be guarded behind a MSVC generator expression

**Review by jsantiago-eProsima — CHANGES_REQUESTED** [review]

File: src/cpp/CMakeLists.txt (hunk 1 above)
Comment (richiware) [reply src/cpp/CMakeLists.txt:63]: It is a CMake generator expression, and always will be resolved in the User's cmake configuration step.  Adding the check Win32 doesn't avoid to have the generator expresion in the generated `fastrtps-...-config.cmake` file. Only when the User's cmake configuration step resolves the generator expression will not put the `-DFASTRTPS_DYN_LINK`, but in linux it doesn't hurt. My opinion is to have the simplier generator expression.

**Review by jsantiago-eProsima — APPROVED** [review]

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/158/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Mac/86/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/150/)

**Pull request merged**